[C99] Claim conformance to WG14 N717

AaronBallman opened the pull request:
This was the paper that added Universal Character Names to C.

AaronBallman committed:
[C99] Claim conformance to WG14 N717

This was the paper that added Universal Character Names to C.

cor3ntin commented on clang/test/C/C99/n717.c:
```diff
@@ -0,0 +1,69 @@
+// A universal character name shall not specify a character whose short
+// identifier is less than 00A0 other than 0024 ($), 0040 (@), or 0060 (�), nor
+// one in the range D800 through DFFF inclusive.
+//
+// We use a python script to generate the test contents for the large ranges
+// without edge cases.
+// RUN: %python %S/n717.py >%t.inc
+// RUN: %clang_cc1 -verify -std=c99 -Wno-unicode-whitespace -Wno-unicode-homoglyph -Wno-unicode-zero-width -Wno-mathematical-notation-identifier-extension %t.inc
+
```
This seem excessive, can we just test at the boundaries at the range?

AaronBallman replied:
Yeah, I was worried about this being a bit excessive (the test takes about 2s to run on my machine with a debug build, so it's a heavier test than usual but not the worst in terms of performance). But it also seems like this is something we can test exhaustively to ensure we don't have special handling for some particular UCN that causes issues (like assertions) because some of these diagnostics are generated by values in the middle of the range rather than at the edges.

That said, if you have a suggestion for a better approach, I'm game!

cor3ntin commented on clang/test/C/C99/n717.c:
```diff
@@ -0,0 +1,69 @@
+// RUN: %clang_cc1 -verify -std=c99 %s
+// RUN: %clang_cc1 -verify -std=c99 -fno-dollars-in-identifiers %s
+
+/* WG14 N717: Clang 17
+ * Extended identifiers
+ */
```
Do you know what was not supported before?

AaronBallman replied:
UCNs weren't supported in C89 but were added in C99, so this is testing UCNs as specified in https://www.open-std.org/jtc1/sc22/wg14/www/docs/n717.htm but based on whatever was in the final text of the C99 standard (so it incorporates changes from NB comments, etc).

cor3ntin replied:
I mean, before clang 17. I don't remember we changed anything there (for C99) in a while.

AaronBallman replied:
Ah! We were missing some diagnostics: https://godbolt.org/z/hEKsEKqzT

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-wg14-n717

AaronBallman committed:
Update based on review comments

This removes the Python script as overkill and adds an additional test
case.

AaronBallman committed:
Update the test based on offline discussions

Corentin pointed out that UCNs are converted in Phase 5 when they're
not part of an identifier, so I can't use my macro trick to test them.

Also added some tests for edge cases. Still claiming full support as of
Clang 17 for this despite there being some rough edges.

AaronBallman committed:
Switch from character constant to string constant

This removes the FIXME, which is better handled via filing an issue.

AaronBallman committed:
Add some more tests based on offline discussions

cor3ntin reviewed (approved):
Thanks Aaron for putting up with all my offline requests. LGTM

AaronBallman commented:
> Thanks Aaron for putting up with all my offline requests. LGTM

Thank you for all the excellent suggestions, this was a lot harder to test correctly than I realized at first.
